go/worker/compute/executor/committee/node.go

```diff
@@ -417,13 +496,244 @@ func (n *Node) HandleNewBlockLocked(blk *block.Block) {
+	// If we are not a leader don't do anything.
+	if !epoch.IsTransactionSchedulerLeader(round + 1) {
+		return errNotSchedulingLeader
+	}
+
+	// Leader node opens a new parent span for batch processing.
+	batchSpan := opentracing.StartSpan("TakeBatchFromQueue(batch)")
+	defer batchSpan.Finish()
+	batchSpanCtx := batchSpan.Context()
+
+	// Generate the initial I/O root containing only the inputs (outputs and
+	// tags will be added later by the executor nodes).
+	lastHeader := n.commonNode.CurrentBlock.Header
+	emptyRoot := storage.Root{
+		Namespace: lastHeader.Namespace,
+		Version:   lastHeader.Round + 1,
+	}
+	emptyRoot.Hash.Empty()
+
+	ioTree := transaction.NewTree(nil, emptyRoot)
+	defer ioTree.Close()
+
+	for idx, tx := range batch {
+		if err := ioTree.AddTransaction(n.ctx, transaction.Transaction{Input: tx, BatchOrder: uint32(idx)}, nil); err != nil {
+			n.logger.Error("failed to create I/O tree",
+				"err", err,
+			)
+			return err
+		}
+	}
+
+	ioWriteLog, ioRoot, err := ioTree.Commit(n.ctx)
+	if err != nil {
+		n.logger.Error("failed to create I/O tree",
+			"err", err,
+		)
+		return err
+	}
+
+	// Commit I/O tree to storage and obtain receipts.
+	spanInsert, ctx := tracing.StartSpanWithContext(n.ctx, "Apply(ioWriteLog)",
+		opentracing.ChildOf(batchSpanCtx),
+	)
+
+	ioReceipts, err := n.commonNode.Storage.Apply(ctx, &storage.ApplyRequest{
+		Namespace: lastHeader.Namespace,
+		SrcRound:  lastHeader.Round + 1,
+		SrcRoot:   emptyRoot.Hash,
+		DstRound:  lastHeader.Round + 1,
+		DstRoot:   ioRoot,
+		WriteLog:  ioWriteLog,
+	})
+	if err != nil {
+		spanInsert.Finish()
+		n.logger.Error("failed to commit I/O tree to storage",
+			"err", err,
+		)
+		return err
+	}
+	spanInsert.Finish()
+
+	// Dispatch batch to group.
+	spanPublish := opentracing.StartSpan("PublishScheduledBatch(batchHash, header)",
+		opentracing.Tag{Key: "ioRoot", Value: ioRoot},
+		opentracing.Tag{Key: "header", Value: n.commonNode.CurrentBlock.Header},
+		opentracing.ChildOf(batchSpanCtx),
+	)
+	ioReceiptSignatures := []signature.Signature{}
+	for _, receipt := range ioReceipts {
+		ioReceiptSignatures = append(ioReceiptSignatures, receipt.Signature)
+	}
+
+	dispatchMsg := &commitment.ProposedBatch{
+		IORoot:            ioRoot,
+		StorageSignatures: ioReceiptSignatures,
+		Header:            n.commonNode.CurrentBlock.Header,
+	}
+	signedDispatchMsg, err := commitment.SignProposedBatch(n.commonNode.Identity.NodeSigner, dispatchMsg)
+	if err != nil {
+		n.logger.Error("failed to sign txn scheduler batch",
+			"err", err,
+		)
+		return fmt.Errorf("failed to sign txn scheduler batch: %w", err)
+	}
+
+	err = n.commonNode.Group.Publish(
+		batchSpanCtx,
+		&p2p.Message{
+			ProposedBatch: signedDispatchMsg,
+		},
+	)
+	if err != nil {
+		spanPublish.Finish()
+		n.logger.Error("failed to publish batch to committee",
+			"err", err,
+		)
+		return err
+	}
+	crash.Here(crashPointLeaderBatchPublishAfter)
+	spanPublish.Finish()
+
+	if epoch.IsExecutorMember() {
```

Comment:
I think this will currently always be true as a transaction scheduler is one of the executors? So it should be safe to remove this check and add a comment saying something like "Also process the batch locally.".